apps/session-manager/src/controllers/fast-login.ts

```diff
@@ -0,0 +1,71 @@
+import {
+  IResponseErrorForbiddenNotAuthorized,
+  IResponseErrorInternal,
+  IResponseSuccessJson,
+  ResponseErrorInternal,
+  ResponseSuccessJson,
+} from "@pagopa/ts-commons/lib/responses";
+import { flow, pipe } from "fp-ts/function";
+import * as E from "fp-ts/lib/Either";
+import * as TE from "fp-ts/lib/TaskEither";
+import * as RTE from "fp-ts/ReaderTaskEither";
+import { readableReportSimplified } from "@pagopa/ts-commons/lib/reporters";
+import { GenerateNonceResponse } from "../generated/fast-login-api/GenerateNonceResponse";
+import { readableProblem } from "../utils/errors";
+import {
+  IResponseErrorUnauthorized,
+  ResponseErrorStatusNotDefinedInSpec,
+  ResponseErrorUnexpectedAuthProblem,
+} from "../utils/responses";
+import { getFastLoginLollipopConsumerClient } from "../repositories/fast-login-api";
+
+export const generateNonceEndpoint: RTE.ReaderTaskEither<
+  {
+    client: ReturnType<getFastLoginLollipopConsumerClient>;
+  },
+  Error,
+  | IResponseErrorUnauthorized
+  | IResponseErrorForbiddenNotAuthorized
+  | IResponseErrorInternal
+  | IResponseSuccessJson<GenerateNonceResponse>
+> = ({ client }) =>
+  TE.tryCatch(
+    async () =>
+      pipe(
+        TE.tryCatch(
+          () => client.generateNonce({}),
+          (_) =>
+            ResponseErrorInternal("Error while calling fast-login service"),
+        ),
+        TE.chainEitherKW(
+          E.mapLeft(
+            flow(readableReportSimplified, (message) =>
+              ResponseErrorInternal(
+                `Unexpected response from fast-login service: ${message}`,
+              ),
+            ),
+          ),
+        ),
+        TE.chainW((response) => {
+          switch (response.status) {
+            case 200:
+              return TE.right(ResponseSuccessJson(response.value));
+            case 401:
+              return TE.left(ResponseErrorUnexpectedAuthProblem());
+            case 500:
+              return TE.left(
+                ResponseErrorInternal(readableProblem(response.value)),
+              );
+            case 502:
+            case 504:
+              return TE.left(
+                ResponseErrorInternal("An error occurred on upstream service"),
+              );
+            default:
+              return TE.left(ResponseErrorStatusNotDefinedInSpec(response));
```

Comment:
If the response is not defined in spec, the executing code should never come here. This check is needed when a spec update comes and the controller code is not updated to handle properly the new status code causing a build error. An `assertNever` should be a better choise.

Reply:
this is already the case. for instance, adding a response status 505 to `GenerateNonceT` type like so:
```ts
// Request type definition
export type GenerateNonceT = r.IPostApiRequestType<
  { readonly ApiKeyAuth: string },
  "X-Functions-Key",
  never,
  | r.IResponseType<200, GenerateNonceResponse, never>
  | r.IResponseType<401, undefined, never>
  | r.IResponseType<500, InternalServerError, never>
  | r.IResponseType<502, undefined, never>
  | r.IResponseType<504, undefined, never>
  // new status code
  | r.IResponseType<505, undefined, never>
>;
```

brings a typescript error on line 65 -> `typescript: Argument of type 'IResponseType<505, undefined, never>' is not assignable to parameter of type 'never'.`

Reply:
right, but you don't need to return a response like `ResponseErrorStatusNotDefinedInSpec` does.

Reply:
what do you think about an assertNever function with a throw inside?